lib/fluent/plugin/out_azure-storage-append-blob.rb

```diff
@@ -17,68 +18,112 @@ class AzureStorageAppendBlobOut < Fluent::Plugin::Output
       config_param :format, :string, :default => DEFAULT_FORMAT_TYPE
-      config_param :time_slice_format, :string, :default => '%Y%m%d'
-  
+      config_param :time_slice_format, :string, :default => "%Y%m%d"
+      config_param :localtime, :bool, :default => false
+
       config_section :format do
         config_set_default :@type, DEFAULT_FORMAT_TYPE
       end
-  
+
       config_section :buffer do
-        config_set_default :chunk_keys, ['time']
+        config_set_default :chunk_keys, ["time"]
         config_set_default :timekey, (60 * 60 * 24)
       end
-  
+
       attr_reader :bs
-  
+
       def configure(conf)
         super
-  
+
         @formatter = formatter_create
-  
+
         if @localtime
-          @path_slicer = Proc.new {|path|
+          @path_slicer = Proc.new { |path|
             Time.now.strftime(path)
           }
         else
-          @path_slicer = Proc.new {|path|
+          @path_slicer = Proc.new { |path|
             Time.now.utc.strftime(path)
           }
         end
-  
+
+        if @azure_storage_account.nil?
+          raise ConfigError, "storage account needs to be specified"
+        end
+
         if @azure_container.nil?
-          raise ConfigError, 'azure_container is needed'
+          raise ConfigError, "container needs to be specified"
+        end
+
+        if !@azure_use_managed_identity && @azure_storage_access_key.nil?
+          raise ConfigError, "either managed identity is enabled or storage account key needs to be present"
         end
       end
-  
+
       def multi_workers_ready?
         true
       end
-  
+
+      def get_access_token()
+        access_key_request = Faraday.new("http://169.254.169.254/metadata/identity/oauth2/token?" +
+                                         "api-version=2019-08-15" +
+                                         "&resource=https://storage.azure.com/",
+                                         headers: { "Metadata" => "true" })
+          .get
+          .body
+        return JSON.parse(access_key_request)["access_token"]
+      end
+
       def start
         super
-  
-        @bs = Azure::Storage::Blob::BlobService.create(storage_account_name: @azure_storage_account, storage_access_key: @azure_storage_access_key)
-  
-        ensure_container
+        if @azure_use_managed_identity
+          token_credential = Azure::Storage::Common::Core::TokenCredential.new get_access_token
+          token_signer = Azure::Storage::Common::Core::Auth::TokenSigner.new token_credential
+          @bs = Azure::Storage::Blob::BlobService.new(storage_account_name: @azure_storage_account, signer: token_signer)
 
-        @azure_storage_path = ''
-        @last_azure_storage_path = ''
+          # Refresh internal is 50 minutes
+          refresh_interval = 50 * 60
+          # The user-defined thread that renews the access token
+          cancelled = false
+          renew_token = Thread.new do
+            Thread.stop
```

Comment:
Sorry for my lack of Ruby experience.. What is this `Thread.stop` for? 
Isn't it immediate thread exit?

Reply:
I'm not a ruby person either, this is taken straight from the official [azure-storage-ruby sdk docs](https://github.com/Azure/azure-storage-ruby/tree/master/blob#access-token)

Reply:
@logachev I debugged this threading code and it works as expected. According to the documentation for [Thread.stop](https://ruby-doc.org/core-2.5.0/Thread.html) it puts the thread to sleep and resumes in the next line which enters the loop. I suspect it's done to unblock the rest of the code right after kicking off the thread.

Reply:
Sounds good. Yeah, I was wondering why you did it this way, but it this is aligned with the SDK -- no further questions